// @narik/custom-validators/esm2020/lib/range/directive.mjs
import { Directive, Input, forwardRef, } from '@angular/core';
import { NG_VALIDATORS, } from '@angular/forms';
import { range } from './validator';
import * as i0 from "@angular/core";
const RANGE_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(() => RangeValidator),
    multi: true,
};
export class RangeValidator {
    ngOnInit() {
        this.validator = range(this.range);
    }
    ngOnChanges(changes) {
        for (const key in changes) {
            if (key === 'range') {
                this.validator = range(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    }
    validate(c) {
        return this.validator(c);
    }
    registerOnValidatorChange(fn) {
        this.onChange = fn;
    }
}
RangeValidator.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: RangeValidator, deps: [], target: i0.ɵɵFactoryTarget.Directive });
RangeValidator.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.0.0", type: RangeValidator, selector: "[range][formControlName],[range][formControl],[range][ngModel]", inputs: { range: "range" }, providers: [RANGE_VALIDATOR], usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: RangeValidator, decorators: [{
            type: Directive,
            args: [{
                    selector: '[range][formControlName],[range][formControl],[range][ngModel]',
                    providers: [RANGE_VALIDATOR],
                }]
        }], propDecorators: { range: [{
                type: Input
            }] } });
