// @narik/custom-validators/esm2020/lib/url/directive.mjs
import { Directive, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { url } from './validator';
import * as i0 from "@angular/core";
const URL_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(() => UrlValidator),
    multi: true
};
export class UrlValidator {
    validate(c) {
        return url(c);
    }
}
UrlValidator.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: UrlValidator, deps: [], target: i0.ɵɵFactoryTarget.Directive });
UrlValidator.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.0.0", type: UrlValidator, selector: "[url][formControlName],[url][formControl],[url][ngModel]", providers: [URL_VALIDATOR], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: UrlValidator, decorators: [{
            type: Directive,
            args: [{
                    selector: '[url][formControlName],[url][formControl],[url][ngModel]',
                    providers: [URL_VALIDATOR]
                }]
        }] });
